### DISCUSSION TITLE
[typescript] Type 'never' has no call signatures

### DISCUSSION THREAD

**Discussion by matallui** [pmndrs/jotai#1308] [Q&A] [answered]
I have this React Native project set up with Typescript and am using Jotai for the first time.
One issue I've encountered so far was that when I pass an atom as a prop to a component, the `useAtom()` call inside that component will return a type of `never` for the setter function.

I've found out that when an atom is initialized with a `null | undefined` value, the setter defaults to `never`.
However, the atoms being passed to the component are all initialized.

Is this an issue that needs to be addressed? Or am I doing something wrong?

Example:

```ts
// atoms.ts
// ...

export const atomA = atom(false);
export const atomB = atom(true);
```

```tsx
// Parent.tsx
import { atomA, atomB } from './atoms.ts';

const Parent = () => {
  return (
    <>
      <Child atom={atomA} />
      <Child atom={atomB} />
    </>
  )
}
```

```tsx
// Child.tsx
import { Atom, useAtom } from 'jotai';

interface Props {
  atom: Atom<boolean>;
}

const Child: React.FC<Props> = ({ atom }) => {
  const [theAtom, setTheAtom] = useAtom(atom);

  // TSC error: setTheAtom: Type 'never' has no call signatures.

  // ...
  return <></>
}
```

**Comment by dai-shi** [comment, marked as answer]
Please try this:
```ts
// Child.tsx
import { PrimitiveAtom, useAtom } from 'jotai';

interface Props {
  atom: PrimitiveAtom<boolean>;
}
```

`Atom<T>` represents read-only atom type, which you can't write.

**Reply by matallui** [reply]
Awesome, thanks! That did it.

A workaround I had in place was to:
```ts
// atoms.ts
export type SomeAtom = typeof atomA;
```

And then use that type instead. And I did notice the `PrimitiveAtom` type there, I guess I just didn't think of trying that.

Thanks again! 🙏🏼

**Reply by dai-shi** [reply]
`typeof atomA` is also a very good solution (not a workaround).

**Reply by damianon** [reply]
How should this case be handled?

```js
import { atom, useAtom } from "jotai";

const inEditAtom = atom<string|null>(null);

export const useEditState = () => {
  const [inEdit, setInEdit] = useAtom(inEditAtom);

  const startEdit = (editing: string) => setInEdit(editing);
  const endEdit = () => setInEdit(null);

  return { inEdit, startEdit, endEdit };
};
```

**Reply by dai-shi** [reply]
Looks just good: https://tsplay.dev/wQ8EJW

**Reply by max-programming** [reply]
@damianon https://github.com/pmndrs/jotai/issues/802#issuecomment-951732589 has the solution. In short, enable `strict` and/or `strictNullChecks` in your `tsconfig.json`
